controller.go

```diff
@@ -54,7 +65,46 @@ func (c *OFController) sendEchoLoop() {
 	// send echo request forever
 }
 
-func ServerLoop(listenPort int) {
+func (c *OFController) Serve(l net.Listener) error {
+	if c.isShuttingDown() {
+		return http.ErrServerClosed
+	}
+
+	c.listener = l
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			if c.isShuttingDown() {
+				return nil
+			}
+			return err
+		}
+		go func() {
+			dp := handleConnection(conn, c)
+			if dp != nil {
+				c.mtx.Lock()
+				c.dps = append(c.dps, dp)
+				dp.RegisterOnClose(c.onDpClosed)
+				c.mtx.Unlock()
+			}
+		}()
+	}
+}
+
+func (c *OFController) onDpClosed(inDp *Datapath) {
+	if c.isShuttingDown() {
```

Comment:
This is an unsafe check (could hit line 98 with `c.isShuttingDown() == true`), so it should be removed. But it is also not clear that the rest is correct. Fix it so that it is clear that there is no recursive `c.mtx.Lock`.

Reply:
Removed. On shutdown deadlock didn't happen.